rasa_nlu/data_router.py

```diff
@@ -4,37 +4,69 @@
 from __future__ import unicode_literals
 
 import datetime
+
 import glob
 import json
 import logging
-import multiprocessing
 import os
 import tempfile
+import io
 
 from builtins import object
 from typing import Text
 
+from concurrent.futures import ProcessPoolExecutor as ProcessPool
+from twisted.internet import defer
+from twisted.logger import jsonFileLogObserver, Logger
+
 from rasa_nlu import utils
 from rasa_nlu.components import ComponentBuilder
 from rasa_nlu.config import RasaNLUConfig
 from rasa_nlu.model import Metadata, InvalidModelError, Interpreter
-from rasa_nlu.train import do_train
+from rasa_nlu.train import do_train_in_worker
 
 logger = logging.getLogger(__name__)
 
 
+def deferred_from_future(future):
+    """Converts a concurrent.futures.Future object to a twisted.internet.defer.Deferred obejct.
+    See: https://twistedmatrix.com/pipermail/twisted-python/2011-January/023296.html
+    """
+    d = defer.Deferred()
+
+    def callback(future):
+        e = future.exception()
+        if e:
+            d.errback(e)
+            return
+        d.callback(future.result())
+
+    future.add_done_callback(callback)
+    return d
+
+
 class DataRouter(object):
     DEFAULT_MODEL_NAME = "default"
 
     def __init__(self, config, component_builder):
+        self._training_processes = config['max_training_processes'] if config['max_training_processes'] > 0 else 1
         self.config = config
         self.responses = DataRouter._create_query_logger(config['response_log'])
-        self._train_procs = []
+        self._trainings_queued = 0
```

Comment:
I know the low-level training process tracking was replaced by this simple counter. But, in my opinion, as the processes are now managed by the `ProcessPool` object, they are too low-level to be accessed and managed directly. This situation could be improved a lot by finishing the transition to a higher-level endpoints tracking (see my other PR).